### ISSUE TITLE
Docker and distillery deployement

### ISSUE THREAD

**Issue by happysalada** [hansihe/rustler#164]
I'm running into an issue deploying with docker and distillery.
Running a local deployment with distillery works fine. However when I do the same deployment with docker then I get this error.
```
web_1         | 2018-07-12 23:43:16 crash_report
web_1         |     initial_call: {supervisor,kernel,['Argument__1']}
web_1         |     pid: <0.2238.0>
web_1         |     registered_name: []
web_1         |     error_info: {exit,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'},[{init,run_on_load_handlers,0,[]},{kernel,init,1,[{file,"kernel.erl"},{line,217}]},{supervisor,init,1,[{file,"supervisor.erl"},{line,294}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,365}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,333}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
web_1         |     ancestors: [kernel_sup,<0.2214.0>]
web_1         |     message_queue_len: 0
web_1         |     messages: []
web_1         |     links: [<0.2215.0>]
web_1         |     dictionary: []
web_1         |     trap_exit: true
web_1         |     status: running
web_1         |     heap_size: 610
web_1         |     stack_size: 27
web_1         |     reductions: 235
web_1         | 2018-07-12 23:43:16 supervisor_report
web_1         |     supervisor: {local,kernel_sup}
web_1         |     errorContext: start_error
web_1         |     reason: {on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}
web_1         |     offender: [{pid,undefined},{id,kernel_safe_sup},{mfargs,{supervisor,start_link,[{local,kernel_safe_sup},kernel,safe]}},{restart_type,permanent},{shutdown,infinity},{child_type,supervisor}]
web_1         | 2018-07-12 23:43:17 crash_report
web_1         |     initial_call: {application_master,init,['Argument__1','Argument__2','Argument__3','Argument__4']}
web_1         |     pid: <0.2213.0>
web_1         |     registered_name: []
web_1         |     error_info: {exit,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{kernel,start,[normal,[]]}},[{application_master,init,4,[{file,"application_master.erl"},{line,134}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
web_1         |     ancestors: [<0.2212.0>]
web_1         |     message_queue_len: 1
web_1         |     messages: [{'EXIT',<0.2214.0>,normal}]
web_1         |     links: [<0.2212.0>,<0.2211.0>]
web_1         |     dictionary: []
web_1         |     trap_exit: true
web_1         |     status: running
web_1         |     heap_size: 376
web_1         |     stack_size: 27
web_1         |     reductions: 213
web_1         | 2018-07-12 23:43:17 std_info
web_1         |     application: kernel
web_1         |     exited: {{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{kernel,start,[normal,[]]}}
web_1         |     type: permanent
web_1         | {"Kernel pid terminated",application_controller,"{application_start_failure,kernel,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{kernel,start,[normal,[]]}}}"}
web_1         | Kernel pid terminated (application_controller) ({application_start_failure,kernel,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{ke
web_1         |
web_1         | Crash dump is being written to: erl_crash.dump...done
```
when I try to start the release with the foreground command

Here is my docker file
```
FROM elixir:1.6.6-alpine

ARG APP_NAME=union
ARG MIX_ENV=prod
ARG APP_VERSION=0.0.0
ARG PHOENIX_SUBDIR=.
ENV MIX_ENV ${MIX_ENV}
ENV APP_VERSION ${APP_VERSION}
ENV REPLACE_OS_VARS true

WORKDIR /opt/app
# use yarn instead of npm to reduce from 180s to 60s
# add git since one dependency pulls from git
RUN apk update \
  && apk --no-cache --update add nodejs yarn rust cargo git build-base \
  && mix local.rebar --force \
  && mix local.hex --force
COPY . .
RUN mix do deps.get, deps.compile, compile

RUN cd ${PHOENIX_SUBDIR}/assets \
  && yarn install \
  && yarn deploy \
  && cd .. \
  && mix phx.digest
RUN mix release --env=${MIX_ENV} --verbose \
  && mv _build/prod/rel/${APP_NAME} /opt/release \
  && mv /opt/release/bin/${APP_NAME} /opt/release/bin/start_server

# minimal runtime image
FROM alpine:3.8
# bash is required by distillery
RUN apk update && apk --no-cache --update add bash openssl-dev
ENV REPLACE_OS_VARS true
WORKDIR /opt/app
COPY --from=0 /opt/release .
CMD ["/opt/app/bin/start_server", "foreground"]
```

I have checked in the docker container that the priv/native contains the .so files

Somebody pointed me to this issue https://github.com/riverrun/argon2_elixir/issues/5
However I have check that smp is enabled though, and dirty schedulers is enabled since otp 19

I'm not really sure where to look next, any clues/pointers would be appreciated.
I understand that this might not be the right place to post this issue. If you feel strongly against it, feel free to close it.

**Comment by scrogson** [comment]
Try adding `musl-dev` to your `apk add` since you are using Alpine.

**Comment by happysalada** [comment]
@scrogson thanks for the advice. Unfortunately I still get the same error

**Comment by bitwalker** [comment]
@happysalada are you sure that `elixir:1.6.6-alpine` and `alpine:3.8` are both Alpine 3.8 images? That would be my first guess as to a potential problem, since compiled artifacts across Alpine versions may not load. The fact that the release will boot within the same container, but not when brought across to another container indicates that this is more likely an issue with something missing in the target container, rather than a build issue per se.

You could try setting `-loader_init` and/or `-init_debug` in your `vm.args` to see if that yields more information during boot about what might be causing the failure.

**Comment by happysalada** [comment]
@bitwalker thanks for this! Indeed they are not the same versions. I was lazily waiting for the next version of distillery to be able to release the next version of elixir which is 3.8. (I'll try dowgrading instead)

thanks also for the `vm.args` tip, I'll give it a try!

**Comment by happysalada** [comment]
That experiment yielded some interesting results.
elixir `1.6.6-alpine` is based on `erlang-20-alpine` which was originally alpine 3.7, but was [bumped to alpine 3.8](https://github.com/erlang/docker-erlang-otp/commit/70745eada41c444f663484114ab942192b47337b#diff-0498404ad1c8f2ef0bfd27330beec05b). If there is a need to stay consistent with the alpine version, everytime a new version comes out, you have to watch out if one of your docker image doesn't get updated without you knowing it.

adding the two flags in vm.args, yielded the following informations
```
| {progress,applications_loaded}
web_1         | {apply,{application,start_boot,[kernel,permanent]}}
web_1         | {running_on_load_handler,'Elixir.Union.ConsolidatedExecution'}
web_1         | {on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}
web_1         | 2018-08-08 01:31:19 crash_report
web_1         |     initial_call: {supervisor,kernel,['Argument__1']}
web_1         |     pid: <0.2243.0>
web_1         |     registered_name: []
web_1         |     error_info: {exit,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'},[{init,run_on_load_handlers,0,[]},{kernel,init,1,[{file,"kernel.erl"},{line,217}]},{supervisor,init,1,[{file,"supervisor.erl"},{line,294}]},{gen_server,init_it,2,[{file,"gen_server.erl"},{line,365}]},{gen_server,init_it,6,[{file,"gen_server.erl"},{line,333}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
web_1         |     ancestors: [kernel_sup,<0.2219.0>]
web_1         |     message_queue_len: 0
web_1         |     messages: []
web_1         |     links: [<0.2220.0>]
web_1         |     dictionary: []
web_1         |     trap_exit: true
web_1         |     status: running
web_1         |     heap_size: 610
web_1         |     stack_size: 27
web_1         |     reductions: 235
web_1         | 2018-08-08 01:31:19 supervisor_report
web_1         |     supervisor: {local,kernel_sup}
web_1         |     errorContext: start_error
web_1         |     reason: {on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}
web_1         |     offender: [{pid,undefined},{id,kernel_safe_sup},{mfargs,{supervisor,start_link,[{local,kernel_safe_sup},kernel,safe]}},{restart_type,permanent},{shutdown,infinity},{child_type,supervisor}]
web_1         | 2018-08-08 01:31:20 crash_report
web_1         |     initial_call: {application_master,init,['Argument__1','Argument__2','Argument__3','Argument__4']}
web_1         |     pid: <0.2218.0>
web_1         |     registered_name: []
web_1         |     error_info: {exit,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{kernel,start,[normal,[]]}},[{application_master,init,4,[{file,"application_master.erl"},{line,134}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
web_1         |     ancestors: [<0.2217.0>]
web_1         |     message_queue_len: 1
web_1         |     messages: [{'EXIT',<0.2219.0>,normal}]
web_1         |     links: [<0.2217.0>,<0.2216.0>]
web_1         |     dictionary: []
web_1         |     trap_exit: true
web_1         |     status: running
web_1         |     heap_size: 376
web_1         |     stack_size: 27
web_1         |     reductions: 213
web_1         | 2018-08-08 01:31:20 std_info
web_1         |     application: kernel
web_1         |     exited: {{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{kernel,start,[normal,[]]}}
web_1         |     type: permanent
web_1         | {apply,{application,start_boot,[stdlib,permanent]}}
web_1         | {"Kernel pid terminated",application_controller,"{application_start_failure,kernel,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{kernel,start,[normal,[]]}}}"}
web_1         | Kernel pid terminated (application_controller) ({application_start_failure,kernel,{{shutdown,{failed_to_start_child,kernel_safe_sup,{on_load_function_failed,'Elixir.Union.ConsolidatedExecution'}}},{ke
web_1         |
web_1         | Crash dump is being written to: erl_crash.dump...done
union_web_1 exited with code 1
```

it seems that the [on_load function is failing](https://github.com/hansihe/rustler/blob/master/rustler_mix/lib/rustler.ex#L43)
looking at it, it is trying to get the nif_path from priv_dir
could it be that the [code for getting a priv_dir](https://github.com/hansihe/rustler/blob/master/rustler_mix/lib/rustler.ex#L76) doesn't work with a release? ie: `otp_app |> :code.priv_dir() |> to_string()`
just a thought.

I feel we are getting closer.
Wondering how to debug this. Is there a way to compile a local version of the library with inspect statement to check what is happening? (just a thought)

**Comment by bitwalker** [comment]
@happysalada The `:code.priv_dir/1` function works in releases (as does `Application.app_dir/2`), and is in fact the recommended way of getting assets in a release, so that isn't the issue.

More likely is that the `.so` file it is trying to load, _can't_ be loaded. This could happen if the compiled artifact references dynamic libraries that are unavailable, not at the paths expected (i.e. not in `LD_LIBRARY_PATH`), or are different versions (i.e. the `.so` expects 1.3.4 of X, but the available version of X is 1.4.0). 

One thing you can try is run `bin/myapp console_clean`, then:

```elixir
iex> Application.load(:myapp)
...> priv_dir = Application.app_dir(:myapp, "priv")
...> nif_path = Path.join([priv_dir, "path", "to", "nif"])
...> load_data = # Whatever this would be set to for this NIF
...> try do 
:erlang.load_nif(String.to_charlist(nif_path), load_data)
catch
  kind, err ->
    {kind, err}
end
```

This will catch the load error and let you see more detail (ideally). You do need to figure out what `load_data` would be for this NIF, but you should be able to work backwards in the Rustler code you linked to find out what that value would be.

**Comment by happysalada** [comment]
I didn't know about the console_clean trick!

Bingo!
```
{:error,
 {:load_failed,
  'Failed to load NIF library: \'Error loading shared library libgcc_s.so.1: No such file or directory (needed by ./lib/union-0.3.3/priv/native/libconsolidated_execution.so)\''}}
```

so adding `apk add build-base` solved that problem

trying that again I ran into
```
:error,
 {:bad_lib,
  'Library module name \'Elixir.Union.ConsolidatedExecution\' does not match calling module \'erl_eval\''}}
```
it seems from [this post](http://blog.differentpla.net/blog/2014/10/09/library-module-does-not-match-calling-module) that you can't call nifs from the shell

so trying it again with a normal start and all boots well!
@bitwalker thanks a lot for the help!
